mplc/multi_partner_learning.py

```diff
@@ -603,12 +604,72 @@ def fit_minibatch(self):
         logger.debug("End of S-Model collaborative round.")
 
 
+class GradientFusion(MultiPartnerLearning):
+    def __init__(self, scenario, **kwargs):
+        super(GradientFusion, self).__init__(scenario, **kwargs)
+        if self.partners_count == 1:
+            raise ValueError('Only one partner is provided. Please use the dedicated SinglePartnerLearning class')
+        self.model = self.build_model()
+
+    def fit_epoch(self):
+        # Split the train dataset in mini-batches
+        self.split_in_minibatches()
+        # Iterate over mini-batches and train
+        for i in range(self.minibatch_count):
+            self.minibatch_index = i
+            self.fit_minibatch()
+
+        self.minibatch_index = 0
+
+    def fit_minibatch(self):
+        """Proceed to a collaborative round with a federated averaging approach"""
+
+        logger.debug("Start new gradients fusion collaborative round ...")
+
+        # Starting model for each partner is the aggregated model from the previous mini-batch iteration
+        logger.info(f"(gradient fusion) Minibatch n°{self.minibatch_index} of epoch n°{self.epoch_index}, "
+                    f"init each partner's models with a copy of the global model")
+
+        for partner in self.partners_list:
+            # Evaluate and store accuracy of mini-batch start model
+            partner.model_weights = self.model_weights
+        self.eval_and_log_model_val_perf()
+
+        # Iterate over partners for training each individual model
+        for partner_index, partner in enumerate(self.partners_list):
+            with tf.GradientTape() as tape:
+                loss = self.model.loss(partner.minibatched_y_train[self.minibatch_index],
+                                       self.model(partner.minibatched_x_train[self.minibatch_index]))
+            partner.grads = tape.gradient(loss, self.model.trainable_weights)
```

Comment:
So in this case it's like `gradient_updates_per_pass_count=1` ?

Reply:
yes